src/tools/pkcs11-tool.c

```diff
@@ -1618,6 +1642,67 @@ static void sign_data(CK_SLOT_ID slot, CK_SESSION_HANDLE session,
 	fprintf(stderr, "Using signature algorithm %s\n", p11_mechanism_to_name(opt_mechanism));
 	memset(&mech, 0, sizeof(mech));
 	mech.mechanism = opt_mechanism;
+	pss_params.hashAlg = 0;
+
+	if (opt_hash_alg != 0 && opt_mechanism != CKM_RSA_PKCS_PSS)
+		util_fatal("The hash-algorithm is applicable only to generic"
+			"RSA-PKCS-PSS mechanism");
+
+	/* set "default" MGF and hash algorithms. We can overwrite MGF later */
+	switch (opt_mechanism) {
+	case CKM_RSA_PKCS_PSS:
+		switch (opt_hash_alg) {
+		case CKM_SHA_1:
+			pss_params.mgf = CKG_MGF1_SHA1;
+			break;
+		case CKM_SHA256:
+			pss_params.mgf = CKG_MGF1_SHA256;
+			break;
+		case CKM_SHA384:
+			pss_params.mgf = CKG_MGF1_SHA384;
+			break;
+		case CKM_SHA512:
+			pss_params.mgf = CKG_MGF1_SHA512;
+			break;
+		default:
+			util_fatal("RSA-PKCS-PSS requires explicit hash mechanism");
```

Comment:
No. Default is using the same hash. E.g., `SHA384-RSA-PKCS-PSS` should use `SHA-384`, `CKG_MGF1_SHA384`, and salt length equal to digest length (aka equivalent to `rsa_pss_saltlen=-1` of OpenSSL).